File: packages/kotlin/ttr-semantics/src/main/kotlin/org/tatrman/ttr/semantics/References.kt
package org.tatrman.ttr.semantics

import org.tatrman.ttr.parser.model.AttributeDef
import org.tatrman.ttr.parser.model.ColumnDef
import org.tatrman.ttr.parser.model.ConstraintDef
import org.tatrman.ttr.parser.model.Definition
import org.tatrman.ttr.parser.model.EntityDef
import org.tatrman.ttr.parser.model.Er2CncRoleDef
import org.tatrman.ttr.parser.model.Er2DbAttributeDef
import org.tatrman.ttr.parser.model.Er2DbEntityDef
import org.tatrman.ttr.parser.model.Er2DbRelationDef
import org.tatrman.ttr.parser.model.FkDef
import org.tatrman.ttr.parser.model.IndexDef
import org.tatrman.ttr.parser.model.ProcedureDef
import org.tatrman.ttr.parser.model.PropertyValue
import org.tatrman.ttr.parser.model.Reference
import org.tatrman.ttr.parser.model.RelationDef
import org.tatrman.ttr.parser.model.SourceLocation
import org.tatrman.ttr.parser.model.TableDef
import org.tatrman.ttr.parser.model.ViewDef

/**
 * One collected cross-reference: the dotted path, its dot-split parts, the source
 * span of the reference token, and the top-level def it belongs to. Mirrors TS
 * `CollectedReference` (`packages/semantics/src/references.ts`).
 *
 * The [Reference] now carries its own [SourceLocation] (matching the TS
 * `Reference`), so both `Reference`-typed slots (nameAttribute, entity, role, …)
 * and `IdValue`-based slots (relation/fk `from`/`to`) report the reference's own
 * span — not the enclosing def's.
 */
data class CollectedRef(
    val path: String,
    val parts: List<String>,
    val source: SourceLocation,
    val ownerDef: Definition,
)

/**
 * Walk every definition (plus its nested children) and collect every
 * cross-reference, paired with the top-level def it came from.
 */
fun collectAllReferences(definitions: List<Definition>): List<CollectedRef> {
    val out = mutableListOf<CollectedRef>()
    for (def in definitions) {
        collectInto(def, def, out)
        for (child in nestedDefs(def)) collectInto(child, def, out)
    }
    return out
}

/** Nested per-def children (attributes, columns, indices, constraints, result columns). */
fun nestedDefs(def: Definition): List<Definition> =
    when (def) {
        is EntityDef -> def.attributes
        is TableDef -> def.columns + def.indices + def.constraints
        is ViewDef -> def.columns
        is ProcedureDef -> def.resultColumns
        else -> emptyList()
    }

private fun collectInto(
    def: Definition,
    owner: Definition,
    out: MutableList<CollectedRef>,
) {
    when (def) {
        is EntityDef -> {
            def.nameAttribute?.let { out += refOf(it, owner) }
            def.codeAttribute?.let { out += refOf(it, owner) }
        }
        is Er2DbEntityDef -> def.entity?.let { out += refOf(it, owner) }
        is Er2DbAttributeDef -> def.attribute?.let { out += refOf(it, owner) }
        is Er2DbRelationDef -> {
            def.relation?.let { out += refOf(it, owner) }
            def.fk?.let { out += refOf(it, owner) }
        }
        is Er2CncRoleDef -> {
            def.entity?.let { out += refOf(it, owner) }
            def.role?.let { out += refOf(it, owner) }
        }
        is RelationDef -> {
            pushIdValue(def.from, owner, out)
            pushIdValue(def.to, owner, out)
        }
        is FkDef -> {
            pushIdValue(def.from, owner, out)
            pushIdValue(def.to, owner, out)
        }
        is AttributeDef, is ColumnDef, is IndexDef, is ConstraintDef -> {}
        else -> {}
    }
}

private fun refOf(
    ref: Reference,
    owner: Definition,
): CollectedRef = CollectedRef(ref.path, ref.parts, ref.source, owner)

private fun pushIdValue(
    value: PropertyValue?,
    owner: Definition,
    out: MutableList<CollectedRef>,
) {
    when (value) {
        is PropertyValue.IdValue -> out += CollectedRef(value.ref.path, value.parts, value.source, owner)
        is PropertyValue.ListValue -> value.items.forEach { pushIdValue(it, owner, out) }
        is PropertyValue.ObjectValue -> value.entries.values.forEach { pushIdValue(it, owner, out) }
        else -> {}
    }
}

private val ENCLOSING_KINDS =
    setOf(
        "entity",
        "table",
        "view",
        "procedure",
        "relation",
        "query",
        "role",
        "er2dbEntity",
        "er2dbAttribute",
        "er2dbRelation",
        "er2cncRole",
    )

/**
 * The qname of the def that lexically encloses a reference (for step-1 lexical
 * resolution). Mirrors TS `enclosingQnameOf` (`reference-index.ts`).
 */
fun enclosingQnameOf(
    def: Definition,
    schemaCode: String,
    namespace: String,
    packageName: String?,
): String? {
    val kind = kindOf(def)
    if (kind !in ENCLOSING_KINDS) return null
    // v4.0 uniform key: model/schema/kind derive from the def's kind; `namespace`
    // is the file `schema` id (db only). `schemaCode` (file model) is unused.
    return buildCanonicalKey(packageName ?: "", namespace, kind, listOf(def.name))
}

File: packages/python/ttr-parser/src/ttr_parser/semantics/references.py
# SPDX-License-Identifier: Apache-2.0
"""Cross-reference collection (← Kotlin `References.kt` / TS `references.ts`).

Walks every definition (and its nested children) and collects every
cross-reference paired with the top-level def it belongs to, plus the helpers
the validator needs: `enclosing_qname_of` (step-1 lexical scope), `kind_of` (the
camelCase kind), and `package_of_import`. Reference-typed slots (nameAttribute,
entity, role, …) and IdValue-based slots (relation/fk from/to) both report the
reference's own source span.
"""

from __future__ import annotations

from dataclasses import dataclass

from ..model import (
    Definition,
    EntityDef,
    Er2CncRoleDef,
    Er2DbAttributeDef,
    Er2DbEntityDef,
    Er2DbRelationDef,
    ExampleDef,
    FkDef,
    IdValue,
    ImportStatement,
    ListValue,
    ObjectValue,
    PatternDef,
    ProcedureDef,
    PropertyValue,
    Reference,
    RelationDef,
    SourceLocation,
    TableDef,
    TermDef,
    ViewDef,
)
from .default_schema import build_canonical_key, kind_segment


@dataclass(frozen=True, slots=True)
class CollectedRef:
    path: str
    parts: tuple[str, ...]
    source: SourceLocation
    owner_def: Definition


def nested_defs(definition: Definition) -> tuple[Definition, ...]:
    """Per-def children: attributes / columns (+ indices + constraints) / result columns."""
    if isinstance(definition, EntityDef):
        return definition.attributes
    if isinstance(definition, TableDef):
        return (*definition.columns, *definition.indices, *definition.constraints)
    if isinstance(definition, ViewDef):
        return definition.columns
    if isinstance(definition, ProcedureDef):
        return definition.result_columns
    return ()


def collect_all_references(definitions: tuple[Definition, ...]) -> list[CollectedRef]:
    out: list[CollectedRef] = []
    for definition in definitions:
        _collect_into(definition, definition, out)
        for child in nested_defs(definition):
            _collect_into(child, definition, out)
    return out


def _ref_of(ref: Reference, owner: Definition) -> CollectedRef:
    return CollectedRef(ref.path, ref.parts, ref.source, owner)


def _collect_into(
    definition: Definition, owner: Definition, out: list[CollectedRef]
) -> None:
    if isinstance(definition, EntityDef):
        if definition.name_attribute is not None:
            out.append(_ref_of(definition.name_attribute, owner))
        if definition.code_attribute is not None:
            out.append(_ref_of(definition.code_attribute, owner))
    elif isinstance(definition, Er2DbEntityDef):
        if definition.entity is not None:
            out.append(_ref_of(definition.entity, owner))
    elif isinstance(definition, Er2DbAttributeDef):
        if definition.attribute is not None:
            out.append(_ref_of(definition.attribute, owner))
    elif isinstance(definition, Er2DbRelationDef):
        if definition.relation is not None:
            out.append(_ref_of(definition.relation, owner))
        if definition.fk is not None:
            out.append(_ref_of(definition.fk, owner))
    elif isinstance(definition, Er2CncRoleDef):
        if definition.entity is not None:
            out.append(_ref_of(definition.entity, owner))
        if definition.role is not None:
            out.append(_ref_of(definition.role, owner))
    elif isinstance(definition, RelationDef):
        _push_id_value(definition.from_, owner, out)
        _push_id_value(definition.to, owner, out)
    elif isinstance(definition, FkDef):
        _push_id_value(definition.from_, owner, out)
        _push_id_value(definition.to, owner, out)
    # v4.4 lexicon entries — the `for:` target ref (er/db/md) resolves through the
    # standard path, giving goto-def + unresolved-reference for free (TS parity).
    elif isinstance(definition, (TermDef, PatternDef, ExampleDef)):
        if definition.target is not None:
            out.append(_ref_of(definition.target, owner))


def _push_id_value(
    value: PropertyValue | None, owner: Definition, out: list[CollectedRef]
) -> None:
    if isinstance(value, IdValue):
        if value.ref is not None:
            out.append(CollectedRef(value.ref.path, value.parts, value.source, owner))
    elif isinstance(value, ListValue):
        for item in value.items:
            _push_id_value(item, owner, out)
    elif isinstance(value, ObjectValue):
        for entry in value.entries.values():
            _push_id_value(entry, owner, out)


_ENCLOSING_KINDS = frozenset(
    {
        "entity",
        "table",
        "view",
        "procedure",
        "relation",
        "query",
        "role",
        "er2dbEntity",
        "er2dbAttribute",
        "er2dbRelation",
        "er2cncRole",
    }
)


def kind_of(definition: Definition) -> str:
    """The camelCase TS kind for a Python `Definition` (e.g. `er2db_entity` → `er2dbEntity`)."""
    return kind_segment(definition.kind)


def enclosing_qname_of(
    definition: Definition,
    schema_code: str,
    namespace: str,
    package_name: str,
) -> str | None:
    """Qname of the def that lexically encloses a reference (step-1 lexical)."""
    kind = kind_of(definition)
    if kind not in _ENCLOSING_KINDS:
        return None
    # v4.0 uniform key: model/schema/kind from the def's kind; `namespace` is the
    # file `schema` id (db only). `schema_code` (file model) is unused.
    return build_canonical_key(
        package_name, namespace, definition.kind, [definition.name]
    )


def package_of_import(imp: ImportStatement) -> str:
    if imp.wildcard:
        return imp.target
    parts = imp.target.split(".")
    return ".".join(parts[:-1]) if len(parts) >= 2 else ""
